stdlib/public/runtime/ImageInspectionStatic.cpp

```diff
@@ -0,0 +1,54 @@
+//===-- ImageInspectionStatic.cpp -------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of functions to read data sections from static executable.
+//
+//===----------------------------------------------------------------------===//
+
+#include "ImageInspection.h"
+
+// Currently only tested on linux but should work for any ELF platform
+#if defined(__ELF__) && defined(__linux__)
+
+// These are defined in swift_sections.S
+// Used in swift_sections.S to mark the start of a section
+struct SectionInfo {
+  uint64_t size;
+  const uint8_t data[0];
```

Comment:
Zero-sized array fields aren't standard C++, and reaching outside the bounds of an array field through a struct pointer is undefined behavior. It'd fit a bit more within the standard to leave `SectionInfo` opaque and do char* pointer manipulation:

```
struct alignas(uint64_t) SectionInfo;
uint64_t getSectionSize(const SectionInfo *section) {
  uint64_t result;
  memcpy(&result, section, sizeof(uint64_t));
  return result;
}
const void *getSectionData(const SectionInfo *section) {
  return reinterpret_cast<const char*>(section) + sizeof(uint64_t);
}
```